# Data preparation

This code is to prepare dataset for modelling a classifier to discriminate emotions using MSP dataset: [Link](https://ecs.utdallas.edu/research/researchlabs/msp-lab/MSP-Podcast.html).

The prerequisite of this notebook is to segment the audio files by following the guidance provided in the MSP dataset. `audio_crop.py` file segmented big chunk audio files into 2-4 audio segments and also 8 seconds in average length audio files for train, test, and validation dataset.

This file is just for practice to extract audio features and put all emotion data into csv files (train, test, validation)

In [1]:
import os
import json
import pandas as pd
from audio_crop import file_labels
import soundfile as sf
import opensmile


In [2]:
# save paths
arouse_path = os.path.relpath('./MSP Data/Annotations/Annotations/Arousal')
dominance_path = os.path.relpath('./MSP Data/Annotations/Annotations/Dominance')
valence_path = os.path.relpath('./MSP Data/Annotations/Annotations/Valence')
train_path = os.path.relpath('./Data/train')
validation_path = os.path.relpath('./Data/validation')
test_path = os.path.relpath('./Data/test')
data_path = os.path.relpath('./Data')

In [3]:
train = pd.DataFrame()
test = pd.DataFrame()
validation = pd.DataFrame()

In [4]:
train_file, validation_files, test_files = file_labels()

Number of training labels: 40
Number of validation labels: 14
Number of testing labels: 20


In [5]:
smile = opensmile.Smile(
    feature_set=opensmile.FeatureSet.eGeMAPSv02,
    feature_level=opensmile.FeatureLevel.Functionals,
    num_workers=4
)


In [116]:
file_paths = []
for file in os.listdir(train_path):
    file_paths.append(os.path.join(train_path, file))
train = smile.process_files(file_paths)

/home/changhyun/workspace/ABI_research/env/lib/python3.8/site-packages/opensmile/core/smile.py:308: UserWarning: Segment too short, filling with NaN.
  warnings.warn(


In [117]:
# Errors in segment.json file that there are the same start and end time for MSP-PODCAST_0153 audio
train[train.isna().any(axis=1)]
# train

F0semitoneFrom27.5Hz_sma3nz_amean  \
file                                 start  end                                         
Data/train/MSP-PODCAST_0153_0162.wav 0 days 0 days                                NaN   
Data/train/MSP-PODCAST_0153_0091.wav 0 days 0 days                                NaN   
Data/train/MSP-PODCAST_0153_0145.wav 0 days 0 days                                NaN   
Data/train/MSP-PODCAST_0153_0136.wav 0 days 0 days                                NaN   
Data/train/MSP-PODCAST_0153_0139.wav 0 days 0 days                                NaN   
...                                                                               ...   
Data/train/MSP-PODCAST_0153_0113.wav 0 days 0 days                                NaN   
Data/train/MSP-PODCAST_0153_0088.wav 0 days 0 days                                NaN   
Data/train/MSP-PODCAST_0153_0097.wav 0 days 0 days                                NaN   
Data/train/MSP-PODCAST_0153_0155.wav 0 days 0 days                                NaN   
Data/train/MSP-PODCAST_0153_0100.wav 0 days 0 days                                NaN   

                                                    F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file                                 start  end                                              
Data/train/MSP-PODCAST_0153_0162.wav 0 days 0 days                                     NaN   
Data/train/MSP-PODCAST_0153_0091.wav 0 days 0 days                                     NaN   
Data/train/MSP-PODCAST_0153_0145.wav 0 days 0 days                                     NaN   
Data/train/MSP-PODCAST_0153_0136.wav 0 days 0 days                                     NaN   
Data/train/MSP-PODCAST_0153_0139.wav 0 days 0 days                                     NaN   
...                                                                                    ...   
Data/train/MSP-PODCAST_0153_0113.wav 0 days 0 days                                     NaN   
Data/train/MSP-PODCAST_0153_0088.wav 0 days 0 days                                     NaN   
Data/train/MSP-PODCAST_0153_0097.wav 0 days 0 days                                     NaN   
Data/train/MSP-PODCAST_0153_0155.wav 0 days 0 days                                     NaN   
Data/train/MSP-PODCAST_0153_0100.wav 0 days 0 days                                     NaN   

                                                    F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file                                 start  end                                                  
Data/train/MSP-PODCAST_0153_0162.wav 0 days 0 days                                         NaN   
Data/train/MSP-PODCAST_0153_0091.wav 0 days 0 days                                         NaN   
Data/train/MSP-PODCAST_0153_0145.wav 0 days 0 days                                         NaN   
Data/train/MSP-PODCAST_0153_0136.wav 0 days 0 days                                         NaN   
Data/train/MSP-PODCAST_0153_0139.wav 0 days 0 days                                         NaN   
...                                                                                        ...   
Data/train/MSP-PODCAST_0153_0113.wav 0 days 0 days                                         NaN   
Data/train/MSP-PODCAST_0153_0088.wav 0 days 0 days                                         NaN   
Data/train/MSP-PODCAST_0153_0097.wav 0 days 0 days                                         NaN   
Data/train/MSP-PODCAST_0153_0155.wav 0 days 0 days                                         NaN   
Data/train/MSP-PODCAST_0153_0100.wav 0 days 0 days                                         NaN   

                                                    F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
file                                 start  end                                                  
Data/train/MSP-PODCAST_0153_0162.wav 0 days 0 days                                         NaN   
Data/train/MSP-PODCAST_0153_0091.wav 0 days 0 days                                         NaN   
Data

In [118]:
train = train.reset_index(level=[1,2])
train

start                       end  \
file                                                                    
Data/train/MSP-PODCAST_0227_0263.wav 0 days    0 days 00:00:10.049000   
Data/train/MSP-PODCAST_0135_0064.wav 0 days 0 days 00:00:02.570812500   
Data/train/MSP-PODCAST_0166_0631.wav 0 days    0 days 00:00:06.403250   
Data/train/MSP-PODCAST_0035_0089.wav 0 days    0 days 00:00:04.120000   
Data/train/MSP-PODCAST_1170_0024.wav 0 days    0 days 00:00:05.470000   
...                                     ...                       ...   
Data/train/MSP-PODCAST_1167_0067.wav 0 days    0 days 00:00:00.290000   
Data/train/MSP-PODCAST_0380_0154.wav 0 days           0 days 00:00:01   
Data/train/MSP-PODCAST_0796_0027.wav 0 days           0 days 00:00:16   
Data/train/MSP-PODCAST_0087_0181.wav 0 days 0 days 00:00:08.592062500   
Data/train/MSP-PODCAST_0227_0310.wav 0 days    0 days 00:00:09.821750   

                                      F0semitoneFrom27.5Hz_sma3nz_amean  \
file                                                                      
Data/train/MSP-PODCAST_0227_0263.wav                          29.599983   
Data/train/MSP-PODCAST_0135_0064.wav                          16.518946   
Data/train/MSP-PODCAST_0166_0631.wav                          31.418877   
Data/train/MSP-PODCAST_0035_0089.wav                          24.457880   
Data/train/MSP-PODCAST_1170_0024.wav                          32.391113   
...                                                                 ...   
Data/train/MSP-PODCAST_1167_0067.wav                          36.062950   
Data/train/MSP-PODCAST_0380_0154.wav                          39.779469   
Data/train/MSP-PODCAST_0796_0027.wav                          32.923893   
Data/train/MSP-PODCAST_0087_0181.wav                          23.316153   
Data/train/MSP-PODCAST_0227_0310.wav                          30.400200   

                                      F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file                                                                           
Data/train/MSP-PODCAST_0227_0263.wav                                0.136860   
Data/train/MSP-PODCAST_0135_0064.wav                                0.082243   
Data/train/MSP-PODCAST_0166_0631.wav                                0.141456   
Data/train/MSP-PODCAST_0035_0089.wav                                0.214500   
Data/train/MSP-PODCAST_1170_0024.wav                                0.090431   
...                                                                      ...   
Data/train/MSP-PODCAST_1167_0067.wav                                0.010799   
Data/train/MSP-PODCAST_0380_0154.wav                                0.174736   
Data/train/MSP-PODCAST_0796_0027.wav                                0.118969   
Data/train/MSP-PODCAST_0087_0181.wav                                0.085794   
Data/train/MSP-PODCAST_0227_0310.wav                                0.368234   

                                      F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file                                                                               
Data/train/MSP-PODCAST_0227_0263.wav                                   26.488470   
Data/train/MSP-PODCAST_0135_0064.wav                                   15.134323   
Data/train/MSP-PODCAST_0166_0631.wav                                   26.754557   
Data/train/MSP-PODCAST_0035_0089.wav                                   18.504398   
Data/train/MSP-PODCAST_1170_0024.wav                                   31.491013   
...                                                                          ...   
Data/train/MSP-PODCAST_1167_0067.wav                                   35.676865   
Data/train/MSP-PODCAST_0380_0154.wav                                   35.140995   
Data/train/MSP-PODCAST_0796_0027.wav                                   31.359953   
Data/train/MSP-PODCAST_0087_0181.wav                                   21.812546   
Data/train/MSP-PODCAST_0227_0310.wav                           

In [119]:
# train.dropna()
train = train.dropna()
train

start                       end  \
file                                                                    
Data/train/MSP-PODCAST_0227_0263.wav 0 days    0 days 00:00:10.049000   
Data/train/MSP-PODCAST_0135_0064.wav 0 days 0 days 00:00:02.570812500   
Data/train/MSP-PODCAST_0166_0631.wav 0 days    0 days 00:00:06.403250   
Data/train/MSP-PODCAST_0035_0089.wav 0 days    0 days 00:00:04.120000   
Data/train/MSP-PODCAST_1170_0024.wav 0 days    0 days 00:00:05.470000   
...                                     ...                       ...   
Data/train/MSP-PODCAST_1167_0067.wav 0 days    0 days 00:00:00.290000   
Data/train/MSP-PODCAST_0380_0154.wav 0 days           0 days 00:00:01   
Data/train/MSP-PODCAST_0796_0027.wav 0 days           0 days 00:00:16   
Data/train/MSP-PODCAST_0087_0181.wav 0 days 0 days 00:00:08.592062500   
Data/train/MSP-PODCAST_0227_0310.wav 0 days    0 days 00:00:09.821750   

                                      F0semitoneFrom27.5Hz_sma3nz_amean  \
file                                                                      
Data/train/MSP-PODCAST_0227_0263.wav                          29.599983   
Data/train/MSP-PODCAST_0135_0064.wav                          16.518946   
Data/train/MSP-PODCAST_0166_0631.wav                          31.418877   
Data/train/MSP-PODCAST_0035_0089.wav                          24.457880   
Data/train/MSP-PODCAST_1170_0024.wav                          32.391113   
...                                                                 ...   
Data/train/MSP-PODCAST_1167_0067.wav                          36.062950   
Data/train/MSP-PODCAST_0380_0154.wav                          39.779469   
Data/train/MSP-PODCAST_0796_0027.wav                          32.923893   
Data/train/MSP-PODCAST_0087_0181.wav                          23.316153   
Data/train/MSP-PODCAST_0227_0310.wav                          30.400200   

                                      F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file                                                                           
Data/train/MSP-PODCAST_0227_0263.wav                                0.136860   
Data/train/MSP-PODCAST_0135_0064.wav                                0.082243   
Data/train/MSP-PODCAST_0166_0631.wav                                0.141456   
Data/train/MSP-PODCAST_0035_0089.wav                                0.214500   
Data/train/MSP-PODCAST_1170_0024.wav                                0.090431   
...                                                                      ...   
Data/train/MSP-PODCAST_1167_0067.wav                                0.010799   
Data/train/MSP-PODCAST_0380_0154.wav                                0.174736   
Data/train/MSP-PODCAST_0796_0027.wav                                0.118969   
Data/train/MSP-PODCAST_0087_0181.wav                                0.085794   
Data/train/MSP-PODCAST_0227_0310.wav                                0.368234   

                                      F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file                                                                               
Data/train/MSP-PODCAST_0227_0263.wav                                   26.488470   
Data/train/MSP-PODCAST_0135_0064.wav                                   15.134323   
Data/train/MSP-PODCAST_0166_0631.wav                                   26.754557   
Data/train/MSP-PODCAST_0035_0089.wav                                   18.504398   
Data/train/MSP-PODCAST_1170_0024.wav                                   31.491013   
...                                                                          ...   
Data/train/MSP-PODCAST_1167_0067.wav                                   35.676865   
Data/train/MSP-PODCAST_0380_0154.wav                                   35.140995   
Data/train/MSP-PODCAST_0796_0027.wav                                   31.359953   
Data/train/MSP-PODCAST_0087_0181.wav                                   21.812546   
Data/train/MSP-PODCAST_0227_0310.wav                           

In [120]:
indices = []
for index in train.index:
    indices.append(index[11:32])

train['podcast'] = indices
train = train.set_index('podcast')
train

/tmp/ipykernel_73631/314578783.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  train['podcast'] = indices


start                       end  \
podcast                                                  
MSP-PODCAST_0227_0263 0 days    0 days 00:00:10.049000   
MSP-PODCAST_0135_0064 0 days 0 days 00:00:02.570812500   
MSP-PODCAST_0166_0631 0 days    0 days 00:00:06.403250   
MSP-PODCAST_0035_0089 0 days    0 days 00:00:04.120000   
MSP-PODCAST_1170_0024 0 days    0 days 00:00:05.470000   
...                      ...                       ...   
MSP-PODCAST_1167_0067 0 days    0 days 00:00:00.290000   
MSP-PODCAST_0380_0154 0 days           0 days 00:00:01   
MSP-PODCAST_0796_0027 0 days           0 days 00:00:16   
MSP-PODCAST_0087_0181 0 days 0 days 00:00:08.592062500   
MSP-PODCAST_0227_0310 0 days    0 days 00:00:09.821750   

                       F0semitoneFrom27.5Hz_sma3nz_amean  \
podcast                                                    
MSP-PODCAST_0227_0263                          29.599983   
MSP-PODCAST_0135_0064                          16.518946   
MSP-PODCAST_0166_0631                          31.418877   
MSP-PODCAST_0035_0089                          24.457880   
MSP-PODCAST_1170_0024                          32.391113   
...                                                  ...   
MSP-PODCAST_1167_0067                          36.062950   
MSP-PODCAST_0380_0154                          39.779469   
MSP-PODCAST_0796_0027                          32.923893   
MSP-PODCAST_0087_0181                          23.316153   
MSP-PODCAST_0227_0310                          30.400200   

                       F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
podcast                                                         
MSP-PODCAST_0227_0263                                0.136860   
MSP-PODCAST_0135_0064                                0.082243   
MSP-PODCAST_0166_0631                                0.141456   
MSP-PODCAST_0035_0089                                0.214500   
MSP-PODCAST_1170_0024                                0.090431   
...                                                       ...   
MSP-PODCAST_1167_0067                                0.010799   
MSP-PODCAST_0380_0154                                0.174736   
MSP-PODCAST_0796_0027                                0.118969   
MSP-PODCAST_0087_0181                                0.085794   
MSP-PODCAST_0227_0310                                0.368234   

                       F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
podcast                                                             
MSP-PODCAST_0227_0263                                   26.488470   
MSP-PODCAST_0135_0064                                   15.134323   
MSP-PODCAST_0166_0631                                   26.754557   
MSP-PODCAST_0035_0089                                   18.504398   
MSP-PODCAST_1170_0024                                   31.491013   
...                                                           ...   
MSP-PODCAST_1167_0067                                   35.676865   
MSP-PODCAST_0380_0154                                   35.140995   
MSP-PODCAST_0796_0027                                   31.359953   
MSP-PODCAST_0087_0181                                   21.812546   
MSP-PODCAST_0227_0310                                   24.272110   

                       F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
podcast                                                             
MSP-PODCAST_0227_0263                                   29.002985   
MSP-PODCAST_0135_0064                                   16.496267   
MSP-PODCAST_0166_0631                                   31.878397   
MSP-PODCAST_0035_0089                                   26.782372   
MSP-PODCAST_1170_0024                                   33.067879   
...                                                           ...   
MSP-PODCAST_1167_0067                                   36.027622   
MSP-PODCAST_0380_0154                                   39.348579   
MSP-PODCAST_0796_0027                                

In [123]:
li = os.listdir(arouse_path)
list1 = []
for file in li:
    if file[0:21] in train_file:
        list1.append(file)
list1.sort()
list1
# li[0][0:21] in test_files


['MSP-Conversation_0021_1_001.csv',
 'MSP-Conversation_0021_1_005.csv',
 'MSP-Conversation_0021_1_006.csv',
 'MSP-Conversation_0021_1_007.csv',
 'MSP-Conversation_0021_1_008.csv',
 'MSP-Conversation_0021_1_009.csv',
 'MSP-Conversation_0021_1_010.csv',
 'MSP-Conversation_0021_2_001.csv',
 'MSP-Conversation_0021_2_005.csv',
 'MSP-Conversation_0021_2_006.csv',
 'MSP-Conversation_0021_2_007.csv',
 'MSP-Conversation_0021_2_008.csv',
 'MSP-Conversation_0021_2_009.csv',
 'MSP-Conversation_0021_3_001.csv',
 'MSP-Conversation_0021_3_005.csv',
 'MSP-Conversation_0021_3_006.csv',
 'MSP-Conversation_0021_3_007.csv',
 'MSP-Conversation_0021_3_008.csv',
 'MSP-Conversation_0021_3_010.csv',
 'MSP-Conversation_0021_3_011.csv',
 'MSP-Conversation_0021_4_001.csv',
 'MSP-Conversation_0021_4_002.csv',
 'MSP-Conversation_0021_4_003.csv',
 'MSP-Conversation_0021_4_005.csv',
 'MSP-Conversation_0021_4_006.csv',
 'MSP-Conversation_0021_4_007.csv',
 'MSP-Conversation_0021_4_008.csv',
 'MSP-Conversation_0021_4_00

In [124]:
arouse_annotation = {}
for elem in list1:
    if elem[0:23] in arouse_annotation.keys():
        arouse_annotation.get(elem[0:23]).append(elem)
    else:
        arouse_annotation[elem[0:23]] = [elem]
arouse_annotation

{'MSP-Conversation_0021_1': ['MSP-Conversation_0021_1_001.csv',
  'MSP-Conversation_0021_1_005.csv',
  'MSP-Conversation_0021_1_006.csv',
  'MSP-Conversation_0021_1_007.csv',
  'MSP-Conversation_0021_1_008.csv',
  'MSP-Conversation_0021_1_009.csv',
  'MSP-Conversation_0021_1_010.csv'],
 'MSP-Conversation_0021_2': ['MSP-Conversation_0021_2_001.csv',
  'MSP-Conversation_0021_2_005.csv',
  'MSP-Conversation_0021_2_006.csv',
  'MSP-Conversation_0021_2_007.csv',
  'MSP-Conversation_0021_2_008.csv',
  'MSP-Conversation_0021_2_009.csv'],
 'MSP-Conversation_0021_3': ['MSP-Conversation_0021_3_001.csv',
  'MSP-Conversation_0021_3_005.csv',
  'MSP-Conversation_0021_3_006.csv',
  'MSP-Conversation_0021_3_007.csv',
  'MSP-Conversation_0021_3_008.csv',
  'MSP-Conversation_0021_3_010.csv',
  'MSP-Conversation_0021_3_011.csv'],
 'MSP-Conversation_0021_4': ['MSP-Conversation_0021_4_001.csv',
  'MSP-Conversation_0021_4_002.csv',
  'MSP-Conversation_0021_4_003.csv',
  'MSP-Conversation_0021_4_005.csv',
 

In [125]:
li2 = os.listdir(valence_path)
list2 = []
for file in li2:
    if file[0:21] in train_file:
        list2.append(file)
list2.sort()
list2

['MSP-Conversation_0021_1_001.csv',
 'MSP-Conversation_0021_1_002.csv',
 'MSP-Conversation_0021_1_003.csv',
 'MSP-Conversation_0021_1_005.csv',
 'MSP-Conversation_0021_1_006.csv',
 'MSP-Conversation_0021_1_007.csv',
 'MSP-Conversation_0021_1_008.csv',
 'MSP-Conversation_0021_1_009.csv',
 'MSP-Conversation_0021_1_010.csv',
 'MSP-Conversation_0021_2_001.csv',
 'MSP-Conversation_0021_2_005.csv',
 'MSP-Conversation_0021_2_006.csv',
 'MSP-Conversation_0021_2_007.csv',
 'MSP-Conversation_0021_2_008.csv',
 'MSP-Conversation_0021_2_009.csv',
 'MSP-Conversation_0021_2_010.csv',
 'MSP-Conversation_0021_3_001.csv',
 'MSP-Conversation_0021_3_002.csv',
 'MSP-Conversation_0021_3_003.csv',
 'MSP-Conversation_0021_3_005.csv',
 'MSP-Conversation_0021_3_006.csv',
 'MSP-Conversation_0021_3_007.csv',
 'MSP-Conversation_0021_3_009.csv',
 'MSP-Conversation_0021_4_001.csv',
 'MSP-Conversation_0021_4_005.csv',
 'MSP-Conversation_0021_4_006.csv',
 'MSP-Conversation_0021_4_007.csv',
 'MSP-Conversation_0021_4_00

In [126]:
li3 = os.listdir(dominance_path)
list3 = []
for file in li3:
    if file[0:21] in train_file:
        list3.append(file)
list3.sort()
list3

['MSP-Conversation_0021_1_001.csv',
 'MSP-Conversation_0021_1_005.csv',
 'MSP-Conversation_0021_1_006.csv',
 'MSP-Conversation_0021_1_007.csv',
 'MSP-Conversation_0021_1_008.csv',
 'MSP-Conversation_0021_1_009.csv',
 'MSP-Conversation_0021_1_010.csv',
 'MSP-Conversation_0021_1_011.csv',
 'MSP-Conversation_0021_2_001.csv',
 'MSP-Conversation_0021_2_002.csv',
 'MSP-Conversation_0021_2_003.csv',
 'MSP-Conversation_0021_2_005.csv',
 'MSP-Conversation_0021_2_006.csv',
 'MSP-Conversation_0021_2_007.csv',
 'MSP-Conversation_0021_2_009.csv',
 'MSP-Conversation_0021_2_010.csv',
 'MSP-Conversation_0021_3_001.csv',
 'MSP-Conversation_0021_3_002.csv',
 'MSP-Conversation_0021_3_003.csv',
 'MSP-Conversation_0021_3_005.csv',
 'MSP-Conversation_0021_3_006.csv',
 'MSP-Conversation_0021_3_007.csv',
 'MSP-Conversation_0021_3_010.csv',
 'MSP-Conversation_0021_4_004.csv',
 'MSP-Conversation_0021_4_005.csv',
 'MSP-Conversation_0021_4_006.csv',
 'MSP-Conversation_0021_4_007.csv',
 'MSP-Conversation_0021_4_00

In [127]:
valence_annotation = {}
dominance_annotation = {}
for elem in list2:
    if elem[0:23] in valence_annotation.keys():
        valence_annotation.get(elem[0:23]).append(elem)
    else:
        valence_annotation[elem[0:23]] = [elem]
for elem in list3:
    if elem[0:23] in dominance_annotation.keys():
        dominance_annotation.get(elem[0:23]).append(elem)
    else:
        dominance_annotation[elem[0:23]] = [elem]
valence_annotation

{'MSP-Conversation_0021_1': ['MSP-Conversation_0021_1_001.csv',
  'MSP-Conversation_0021_1_002.csv',
  'MSP-Conversation_0021_1_003.csv',
  'MSP-Conversation_0021_1_005.csv',
  'MSP-Conversation_0021_1_006.csv',
  'MSP-Conversation_0021_1_007.csv',
  'MSP-Conversation_0021_1_008.csv',
  'MSP-Conversation_0021_1_009.csv',
  'MSP-Conversation_0021_1_010.csv'],
 'MSP-Conversation_0021_2': ['MSP-Conversation_0021_2_001.csv',
  'MSP-Conversation_0021_2_005.csv',
  'MSP-Conversation_0021_2_006.csv',
  'MSP-Conversation_0021_2_007.csv',
  'MSP-Conversation_0021_2_008.csv',
  'MSP-Conversation_0021_2_009.csv',
  'MSP-Conversation_0021_2_010.csv'],
 'MSP-Conversation_0021_3': ['MSP-Conversation_0021_3_001.csv',
  'MSP-Conversation_0021_3_002.csv',
  'MSP-Conversation_0021_3_003.csv',
  'MSP-Conversation_0021_3_005.csv',
  'MSP-Conversation_0021_3_006.csv',
  'MSP-Conversation_0021_3_007.csv',
  'MSP-Conversation_0021_3_009.csv'],
 'MSP-Conversation_0021_4': ['MSP-Conversation_0021_4_001.csv',
 

In [128]:
dominance_annotation

{'MSP-Conversation_0021_1': ['MSP-Conversation_0021_1_001.csv',
  'MSP-Conversation_0021_1_005.csv',
  'MSP-Conversation_0021_1_006.csv',
  'MSP-Conversation_0021_1_007.csv',
  'MSP-Conversation_0021_1_008.csv',
  'MSP-Conversation_0021_1_009.csv',
  'MSP-Conversation_0021_1_010.csv',
  'MSP-Conversation_0021_1_011.csv'],
 'MSP-Conversation_0021_2': ['MSP-Conversation_0021_2_001.csv',
  'MSP-Conversation_0021_2_002.csv',
  'MSP-Conversation_0021_2_003.csv',
  'MSP-Conversation_0021_2_005.csv',
  'MSP-Conversation_0021_2_006.csv',
  'MSP-Conversation_0021_2_007.csv',
  'MSP-Conversation_0021_2_009.csv',
  'MSP-Conversation_0021_2_010.csv'],
 'MSP-Conversation_0021_3': ['MSP-Conversation_0021_3_001.csv',
  'MSP-Conversation_0021_3_002.csv',
  'MSP-Conversation_0021_3_003.csv',
  'MSP-Conversation_0021_3_005.csv',
  'MSP-Conversation_0021_3_006.csv',
  'MSP-Conversation_0021_3_007.csv',
  'MSP-Conversation_0021_3_010.csv'],
 'MSP-Conversation_0021_4': ['MSP-Conversation_0021_4_004.csv',
 

In [129]:
segment_path = os.path.relpath('./MSP Data/Time Labels/segments.json')
f = open(segment_path, 'r')
timing_data = json.load(f)
temp = train.copy()

In [130]:
def mean(list):
    return sum(list) / len(list)

In [131]:
for key in timing_data:
    conv_part = timing_data[key]['Conversation_Part']
    if key not in temp.index:
        continue
    start = timing_data[key]['Start_Time']
    end = timing_data[key]['End_Time']
    arousal = []
    valence = []
    dominance = []
    a_annotations = arouse_annotation.get(conv_part)
    v_annotations = valence_annotation.get(conv_part)
    d_annotations = dominance_annotation.get(conv_part)
    for anno in a_annotations:
        pp = os.path.join(arouse_path, anno)
        df = pd.read_csv(pp, header=8, names=['time', 'arousal'])
        df = df[(start <= df['time']) & (df['time'] <= end)]
        if df.empty:
            continue
        arousal.append(df['arousal'].mean())
    
    for anno in v_annotations:
        pp = os.path.join(valence_path, anno)
        df = pd.read_csv(pp, header=8, names=['time', 'valence'])
        df = df[(start <= df['time']) & (df['time'] <= end)]
        if df.empty:
            continue
        valence.append(df['valence'].mean())
    for anno in d_annotations:
        pp = os.path.join(dominance_path, anno)
        df = pd.read_csv(pp, header=8, names=['time', 'dominance'])
        if(anno == 'MSP-Conversation_0047_2_001.csv'):
            df = df.reset_index()
            df = df.drop(columns=['dominance'])
            df.columns = ['time', 'dominance']
        df = df[(start <= df['time']) & (df['time'] <= end)]
        if df.empty:
            continue
        
        dominance.append(df['dominance'].mean())
    
    score_arousal = mean(arousal)
    score_valence = mean(valence)
    score_dominance = mean(dominance)
    
    temp.loc[key, 'arousal'] = score_arousal
    temp.loc[key, 'valence'] = score_valence
    temp.loc[key, 'dominance'] = score_dominance
    
temp

start                       end  \
podcast                                                  
MSP-PODCAST_0227_0263 0 days    0 days 00:00:10.049000   
MSP-PODCAST_0135_0064 0 days 0 days 00:00:02.570812500   
MSP-PODCAST_0166_0631 0 days    0 days 00:00:06.403250   
MSP-PODCAST_0035_0089 0 days    0 days 00:00:04.120000   
MSP-PODCAST_1170_0024 0 days    0 days 00:00:05.470000   
...                      ...                       ...   
MSP-PODCAST_1167_0067 0 days    0 days 00:00:00.290000   
MSP-PODCAST_0380_0154 0 days           0 days 00:00:01   
MSP-PODCAST_0796_0027 0 days           0 days 00:00:16   
MSP-PODCAST_0087_0181 0 days 0 days 00:00:08.592062500   
MSP-PODCAST_0227_0310 0 days    0 days 00:00:09.821750   

                       F0semitoneFrom27.5Hz_sma3nz_amean  \
podcast                                                    
MSP-PODCAST_0227_0263                          29.599983   
MSP-PODCAST_0135_0064                          16.518946   
MSP-PODCAST_0166_0631                          31.418877   
MSP-PODCAST_0035_0089                          24.457880   
MSP-PODCAST_1170_0024                          32.391113   
...                                                  ...   
MSP-PODCAST_1167_0067                          36.062950   
MSP-PODCAST_0380_0154                          39.779469   
MSP-PODCAST_0796_0027                          32.923893   
MSP-PODCAST_0087_0181                          23.316153   
MSP-PODCAST_0227_0310                          30.400200   

                       F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
podcast                                                         
MSP-PODCAST_0227_0263                                0.136860   
MSP-PODCAST_0135_0064                                0.082243   
MSP-PODCAST_0166_0631                                0.141456   
MSP-PODCAST_0035_0089                                0.214500   
MSP-PODCAST_1170_0024                                0.090431   
...                                                       ...   
MSP-PODCAST_1167_0067                                0.010799   
MSP-PODCAST_0380_0154                                0.174736   
MSP-PODCAST_0796_0027                                0.118969   
MSP-PODCAST_0087_0181                                0.085794   
MSP-PODCAST_0227_0310                                0.368234   

                       F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
podcast                                                             
MSP-PODCAST_0227_0263                                   26.488470   
MSP-PODCAST_0135_0064                                   15.134323   
MSP-PODCAST_0166_0631                                   26.754557   
MSP-PODCAST_0035_0089                                   18.504398   
MSP-PODCAST_1170_0024                                   31.491013   
...                                                           ...   
MSP-PODCAST_1167_0067                                   35.676865   
MSP-PODCAST_0380_0154                                   35.140995   
MSP-PODCAST_0796_0027                                   31.359953   
MSP-PODCAST_0087_0181                                   21.812546   
MSP-PODCAST_0227_0310                                   24.272110   

                       F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
podcast                                                             
MSP-PODCAST_0227_0263                                   29.002985   
MSP-PODCAST_0135_0064                                   16.496267   
MSP-PODCAST_0166_0631                                   31.878397   
MSP-PODCAST_0035_0089                                   26.782372   
MSP-PODCAST_1170_0024                                   33.067879   
...                                                           ...   
MSP-PODCAST_1167_0067                                   36.027622   
MSP-PODCAST_0380_0154                                   39.348579   
MSP-PODCAST_0796_0027                                

In [133]:
temp[temp.isnull().any(axis=1)]

Empty DataFrame
Columns: [start, end, F0semitoneFrom27.5Hz_sma3nz_amean, F0semitoneFrom27.5Hz_sma3nz_stddevNorm, F0semitoneFrom27.5Hz_sma3nz_percentile20.0, F0semitoneFrom27.5Hz_sma3nz_percentile50.0, F0semitoneFrom27.5Hz_sma3nz_percentile80.0, F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2, F0semitoneFrom27.5Hz_sma3nz_meanRisingSlope, F0semitoneFrom27.5Hz_sma3nz_stddevRisingSlope, F0semitoneFrom27.5Hz_sma3nz_meanFallingSlope, F0semitoneFrom27.5Hz_sma3nz_stddevFallingSlope, loudness_sma3_amean, loudness_sma3_stddevNorm, loudness_sma3_percentile20.0, loudness_sma3_percentile50.0, loudness_sma3_percentile80.0, loudness_sma3_pctlrange0-2, loudness_sma3_meanRisingSlope, loudness_sma3_stddevRisingSlope, loudness_sma3_meanFallingSlope, loudness_sma3_stddevFallingSlope, spectralFlux_sma3_amean, spectralFlux_sma3_stddevNorm, mfcc1_sma3_amean, mfcc1_sma3_stddevNorm, mfcc2_sma3_amean, mfcc2_sma3_stddevNorm, mfcc3_sma3_amean, mfcc3_sma3_stddevNorm, mfcc4_sma3_amean, mfcc4_sma3_stddevNorm, jitterLocal_sma3nz_amean, jitterLocal_sma3nz_stddevNorm, shimmerLocaldB_sma3nz_amean, shimmerLocaldB_sma3nz_stddevNorm, HNRdBACF_sma3nz_amean, HNRdBACF_sma3nz_stddevNorm, logRelF0-H1-H2_sma3nz_amean, logRelF0-H1-H2_sma3nz_stddevNorm, logRelF0-H1-A3_sma3nz_amean, logRelF0-H1-A3_sma3nz_stddevNorm, F1frequency_sma3nz_amean, F1frequency_sma3nz_stddevNorm, F1bandwidth_sma3nz_amean, F1bandwidth_sma3nz_stddevNorm, F1amplitudeLogRelF0_sma3nz_amean, F1amplitudeLogRelF0_sma3nz_stddevNorm, F2frequency_sma3nz_amean, F2frequency_sma3nz_stddevNorm, F2bandwidth_sma3nz_amean, F2bandwidth_sma3nz_stddevNorm, F2amplitudeLogRelF0_sma3nz_amean, F2amplitudeLogRelF0_sma3nz_stddevNorm, F3frequency_sma3nz_amean, F3frequency_sma3nz_stddevNorm, F3bandwidth_sma3nz_amean, F3bandwidth_sma3nz_stddevNorm, F3amplitudeLogRelF0_sma3nz_amean, F3amplitudeLogRelF0_sma3nz_stddevNorm, alphaRatioV_sma3nz_amean, alphaRatioV_sma3nz_stddevNorm, hammarbergIndexV_sma3nz_amean, hammarbergIndexV_sma3nz_stddevNorm, slopeV0-500_sma3nz_amean, slopeV0-500_sma3nz_stddevNorm, slopeV500-1500_sma3nz_amean, slopeV500-1500_sma3nz_stddevNorm, spectralFluxV_sma3nz_amean, spectralFluxV_sma3nz_stddevNorm, mfcc1V_sma3nz_amean, mfcc1V_sma3nz_stddevNorm, mfcc2V_sma3nz_amean, mfcc2V_sma3nz_stddevNorm, mfcc3V_sma3nz_amean, mfcc3V_sma3nz_stddevNorm, mfcc4V_sma3nz_amean, mfcc4V_sma3nz_stddevNorm, alphaRatioUV_sma3nz_amean, hammarbergIndexUV_sma3nz_amean, slopeUV0-500_sma3nz_amean, slopeUV500-1500_sma3nz_amean, spectralFluxUV_sma3nz_amean, loudnessPeaksPerSec, VoicedSegmentsPerSec, MeanVoicedSegmentLengthSec, StddevVoicedSegmentLengthSec, MeanUnvoicedSegmentLength, StddevUnvoicedSegmentLength, equivalentSoundLevel_dBp, arousal, valence, dominance]
Index: []

[0 rows x 93 columns]

In [100]:
# temp.to_csv('test.csv')

In [115]:
# These values are the total number of each sample
print('Total: ', len(timing_data))
print('Train: ', len(os.listdir(train_path)))
print('Validation: ', len(os.listdir(validation_path)))
print('Test: ', len(os.listdir(test_path)))

Total:  5036
Train:  2783
Validation:  750
Test:  1503


However, there are missing values that openSmile cannot fetch acoustic features from certain audio segments:

Validation dataset: MSP-PODCAST_1188_0020 (Segment name) (Total: 1)
Train dataset: 81 segments in MSP-Conversation_0.153_2.wav were missing (Total: 81) => In segments.json, start time and end time are exactly the same which is 0-length audio segments.

Extraction results:
- Train: 2702
- Validation: 749
- Test: 1503


In [134]:
temp.shape

(2702, 93)